compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -801,6 +809,35 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         }
     }
 
+    // Suggest borrowing the type
+    fn suggest_borrowing_for_object_cast(
+        &self,
+        err: &mut Diagnostic,
+        obligation: &PredicateObligation<'tcx>,
+        self_ty: Ty<'tcx>,
+        object_ty: Ty<'tcx>,
+    ) {
+        let ty::Dynamic(predicates, _) = object_ty.kind() else { return; };
+        let self_ref_ty = self.tcx.mk_imm_ref(self.tcx.lifetimes.re_erased, self_ty);
+
+        for predicate in predicates.iter() {
+            if !self.predicate_must_hold_modulo_regions(
+                &obligation.with(predicate.with_self_ty(self.tcx, self_ref_ty)),
+            ) {
+                return;
+            }
+        }
```

Comment:
Am I correct in my understanding that this is effectively checking “is self conversible to `object`?” Are there any actual examples where you might end up with `ObligationCauseCode::ObjectCastObligation` but where the answer to that question is `false`?

Reply:
This is checking if `&Self` also satisfies the predicate of object, because `Self` is not sized. It may be the case that `T: !Sized` implements `Trait` but `&T` does not.